File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/ClassMemberContainerDetailPane.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail;

import org.gjt.jclasslib.browser.BrowserServices;
import org.gjt.jclasslib.structures.ClassFile;
import org.gjt.jclasslib.structures.ClassMember;
import org.gjt.jclasslib.util.ExtendedJLabel;

import javax.swing.tree.TreePath;

/**
 * Detail pane showing aggregated information about class members.
 *
 * @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
 *
 */
public class ClassMemberContainerDetailPane extends FixedListWithSignatureDetailPane {

    private ExtendedJLabel lblMemberCount;

    /**
     * Constructor.
     *
     * @param services the associated browser services.
     */
    public ClassMemberContainerDetailPane(BrowserServices services, SignatureMode signatureMode) {
        super(services, signatureMode);
    }

    @Override
    protected String getSignatureVerbose() {
        SignatureMode signatureMode = getSignatureMode();
        ClassFile classFile = getServices().getClassFile();
        ClassMember[] classMembers = signatureMode.getClassMembers(classFile);

        StringBuilder buffer = new StringBuilder();
        for (ClassMember classMember : classMembers) {
            signatureMode.appendSignature(classMember, buffer);
            buffer.append('\n');
        }

        return buffer.toString();
    }

    protected void addLabels() {

        addDetailPaneEntry(normalLabel("Member count:"),
                lblMemberCount = highlightLabel());


    }

    public void show(TreePath treePath) {

        ClassFile classFile = getServices().getClassFile();
        int count = getSignatureMode().getClassMembers(classFile).length;
        lblMemberCount.setText(count);

        super.show(treePath);
    }

    protected String getSignatureButtonText() {
        return "Copy signatures to clipboard";
    }
}

File: modules/browser/src/main/kotlin/org/gjt/jclasslib/browser/detail/ClassMemberContainerDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.browser.detail

import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.browser.BrowserTreeNode
import org.gjt.jclasslib.structures.AccessFlag
import org.gjt.jclasslib.structures.ClassMember
import java.awt.GridBagConstraints
import javax.swing.JTree
import javax.swing.tree.TreePath

class ClassMemberContainerDetailPane(services: BrowserServices, signatureMode: FixedListWithSignatureDetailPane.SignatureMode) :
        FixedListWithSignatureDetailPane<Array<out ClassMember>>(Array<out ClassMember>::class.java, services, signatureMode) {

    val filterPane = object : FilterPane<AccessFlag, ClassMember>(this@ClassMemberContainerDetailPane) {
        override fun getAllFilterKeys() = signatureMode.getAccessFlags()
        override fun isElementTextFiltered(element: ClassMember, filterText: String) = isShowAll || element.name.contains(filterText)
        override fun getFilterKeys(element: ClassMember) = AccessFlag.decompose(element.accessFlags)
    }

    override fun show(treePath: TreePath) {
        super.show(treePath)
        updateFilter()
    }

    override fun updateFilter(tree: JTree, treeNode: BrowserTreeNode) {
        super.updateFilter(tree, treeNode)
        filterPane.updateFilterCheckboxes(getElement(TreePath(treeNode.path)).toList())
    }

    override fun isChildShown(node: BrowserTreeNode) = filterPane.isElementShown(node.element as ClassMember)

    override val signatureVerbose: String
        get() = StringBuilder().apply {
            for (classMember in element?.filter { filterPane.isElementShown(it) } ?: listOf()) {
                appendSignature(classMember, signatureMode)
                append('\n')
            }

        }.toString()

    override fun addLabels() {
        add(filterPane, gc() {
            gridx = 0
            gridwidth = GridBagConstraints.REMAINDER
            weightx = 1.0
            fill = GridBagConstraints.HORIZONTAL
        })
        nextLine()

        //TODO add separator
        addDetail("Member count:") { classMembers -> classMembers.count { filterPane.isElementShown(it) }.toString() }
        super.addLabels()
    }

    override val signatureButtonText: String
        get() = "Copy signatures to clipboard"
}

